src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -29,6 +31,19 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// </summary>
     internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, LSP.CompletionList?>
     {
+        /// <summary>
+        /// This sets the max list size we will return in response to a completion request.
+        /// If there are more than this many items, we will set the isIncomplete flag on the returned completion list.
+        /// </summary>
+        private const int MaxCompletionListSize = 1000;
+
+        /// <summary>
+        /// This id is set when we return an incomplete completion list.
+        /// When the client asks us for more items  using <see cref="LSP.CompletionTriggerKind.TriggerForIncompleteCompletions"/>
+        /// we use this result id to lookup the previously calculated list.
+        /// </summary>
+        private long? _lastIncompleteResultId;
```

Comment:
good point, I guess in theory it should, completion requests are not mutating so we could get multiple requests at once.  I wouldn't expect it to usually happen, but I'll do it to be safe

Reply:
Ya one interesting implication of this is if the client sends a completion request and then cancels it this could potentially reference a bad completion list. May want to build in a path to handle / invalidate the "cancelled" state. You'll also want to fallback gracefully in the case that you can't find this resultId too (I see below you currently contract.fail).

Reply:
i assume we don't need a lock as LSP is not concurrent right?  (or is that an incorrect assumption on my part?)
however, we do need to ensure that this is properly cleaned up and is only reused when we are certain it is safe to do so.

Reply:
> i assume we don't need a lock as LSP is not concurrent right? (or is that an incorrect assumption on my part?)

textDocument/didChange requests are not concurrent (they can't be); however, everything else can be. It's up to the client on how it wants to handle that. Also, in practice StreamJsonRpc will only dispatch 1 request at a time, it's up to how you all re-dispatch those requests internally.

Reply:
Mutating requests are not concurrent, but non-mutating requests (this one) could be.  I wouldn't expect to get multiple completion requests though, as generally a completion request is always preceded by a mutating didChange

Reply:
So after looking at this for a while, I think I need to mark completion requests as mutating on our side (which means they're always executed in order and requests wait for the previous requests to finish).  I do not see a way to have a stateless cache for previous incomplete results:
1.  I have no resultId I can pass between the original isIncomplete request and the client request for TriggerForIncompleteCompletion.
2.  Without passing data to the client, I would have to cache on something like document URI + location of the word start character.  But I have no easy way to invalidate that cache (since the cache needs to persist when the document version changes).

thoughts @CyrusNajmabadi @NTaylorMullen

Reply:
Actually even if I do enforce sequential ordering of completion requests, I think we could still have problems.
Imagine the following
```
void M1()
{
    B
}

void M2()
{
    C
}
```

Typing 'B' in M1 gave back an incomplete list, then the user typed 'C' in M2, also giving back an incomplete list.  If the user goes back to 'B' in M1 and types 'o', the client will request for incomplete completions, but we'll use the resultId from 'C'.

Reply:
So I think I have something mostly working -
1.  I now rewind the source text to what it was before completion started (aka delete the filter text).  If the resultId is missing,I use this to re-calculate.
2.  I'm now storing the (rewound) source text with the resultId.  Then when I attempt to use a cached list, I compare the rewound source text to the one stored with the last resultId.  I only re-use the completion list if it matches.
3.  Add a semaphore around reading and writing from the last resultId.  Since we're more stateless now we should be OK as long as we allow only one access to the last resultId at a time.
4.  Also added a test for the code above to verify that it failed on the old version and now passes.

LMK what you all think.

Reply:
> LMK what you all think. In my opinion it would really simplify things if we could pass a resultId to the client for the completion list and have them pass it back to us when they re-request for the incomplete items (similar to how it works with resolve)

That's quite an interesting idea. @dbaeumer have you seen similar scenarios with concurrent language servers trying to respect the isIncomplete -> TriggerForIncompleteCompletion, if have they taken a similar approach to what David did above?

For the record @dibarbet I think what you landed on is great.